File: samples/ocr_e2e/text_rec.py
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.text_rec import TextRecognizer
import common.utils as utils
import numpy as np
import cv2
import argparse
import vaststreamx as vsx


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/resnet34_vd-int8-max-1_3_32_100-vacc/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/crnn_rgbplanar.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--label_file",
        default="../data/labels/key_37.txt",
        help="label file",
    )
    parser.add_argument(
        "--input_file",
        default="../data/images/word_336.png",
        help="input file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="dataset filelist",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        help="dataset output file",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    model = TextRecognizer(
        args.model_prefix,
        args.vdsp_params,
        args.label_file,
        batch_size,
        args.device_id,
        args.hw_config,
    )
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Failed to read input file: {args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        res = model.process(vsx_image)
        print(res)
    else:
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = f.readlines()
        with open(args.dataset_output_file, "wt") as outfile:
            for filename in filelist:
                fullname = os.path.join(args.dataset_root, filename.replace("\n", ""))
                print("fullname:", fullname)
                cv_image = cv2.imread(fullname)
                assert cv_image is not None, f"Read image failed:{filename}"
                vsx_image = utils.cv_bgr888_to_vsximage(
                    cv_image, image_format, args.device_id
                )
                result = model.process(vsx_image)
                basename, _ = os.path.splitext(os.path.basename(fullname))
                result_str = result[0][0]
                outfile.write(f"{basename} {result_str}\n")
        outfile.close()

File: samples/ppocr/text_rec.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/text_rec.hpp"

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "resnet34_vd-int8-max-1_3_32_100-vacc/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/crnn_rgbplanar.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("label_file", '\0', "label file", false,
                        "../data/labels/key_37.txt");
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/word_336.png");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;
  vsx::TextRecognizer text_rec(args.get<std::string>("model_prefix"),
                               args.get<std::string>("vdsp_params"), batch_size,
                               args.get<uint32_t>("device_id"),
                               args.get<std::string>("label_file"),
                               args.get<std::string>("hw_config"));
  auto image_format = text_rec.GetFusionOpIimageFormat();

  if (!args.get<std::string>("dataset_filelist").empty()) {
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    auto dataset_root = args.get<std::string>("dataset_root");
    std::ofstream outfile(args.get<std::string>("dataset_output_file"),
                          std::ios::out);
    CHECK(outfile.is_open())
        << "Failed to open " << args.get<std::string>("dataset_output_file");

    for (size_t s = 0; s < filelist.size(); s++) {
      auto fullname = filelist[s];
      if (!dataset_root.empty()) fullname = dataset_root + "/" + fullname;
      std::cout << fullname << std::endl;
      vsx::Image vsx_image;
      vsx::MakeVsxImage(fullname, vsx_image, image_format);
      auto result = text_rec.Process(vsx_image);
      std::filesystem::path p(fullname);
      std::string basename = p.stem().string();
      auto result_str = vsx::GetStringFromTensor(result);
      outfile << basename << " " << result_str << std::endl;
    }
    outfile.close();

  } else {
    vsx::Image vsx_image;
    vsx::MakeVsxImage(args.get<std::string>("input_file"), vsx_image,
                      image_format);
    auto result = text_rec.Process(vsx_image);
    std::cout << "score: " << vsx::GetScoreFromTensor(result) << std::endl;
    std::cout << "text: " << vsx::GetStringFromTensor(result) << std::endl;
  }

  return 0;
}